Adding a shared device test suite

mariaschuld opened the pull request:
**Context:**

Adds device tests to the ``plugins`` folder. These tests can be applied to any device in the PennyLane ecosystem, to verify that it integrates with PennyLane's core, for example when used to create qnodes and run circuits.

The tests are designed to be overly explicit at the expense of code duplication. Every test should be self-explanatory and take minimum effort to run outside of pytest. Since tests only apply to specific types of devices, the ``_capabilities`` are used to skip tests if necessary.

The test suite will grow bigger in future, this PR introduces the basic skeleton.

**Description of the Change:**

Calling the tests should be done with command line via 

  >>> pytest pennylane/plugins/tests --device default.qubit --shots 60000 --analytic False

The tests are skipped if they do not apply. Currently there are only tests for basic qubit devices (no passthru, no tokens or api, no keyword arguments or backends specified). 

The tests support analytic or non-analytic mode (and hence also hardware). For both, the final probability is computed (to not worry about phase factors introduced by devices), and the tolerance of the assert statement is lowered for non-analytic devices. Flaky is used in addition to increase chances of stochastic tests to pass.  

Note: Also an analytic device has stochastic tests, namely for the ``sample`` return method.

**What is still missing**

It will be a great challenge to extend the tests further, because the devices need to be "cleaned up" and standarized alongside with the growing test suite. Often, this involves design decisions with ADRs -  for example to determine how ``capabilities`` is used in the tests.

Separate CV tests will need to be added.

**Important in this PR**

This PR proposes a few design solutions that should be reviewed before writing more tests:

* Use of flaky
* Use of different tolerance but same tests for analytic and non-analytic devices
* Creating device in fixture factory and passing the function to the tests
* Skipping functions with a simple fixture `skip_if`
* Putting "measurement" and "gate" tests into one file each
* Drawing the command line argument using pytest - **this could still become a problem if we need to allow passing arbitrary arguments**

**EDIT: We decided to implement the test suite for qubit devices only at this stage.**

mariaschuld commented:
I am not sure how to handle CodeFactor and Codecov - obviously the style and coverage of the tests inside the pennylane code folder is not applicable...

mariaschuld commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,23 @@
+* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 
+  to run generic tests on core or external devices calling 
+  
+  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                     >
```
```suggestion
  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                    
```

mariaschuld commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,23 @@
+   
+  * If `--device` is not given, the tests are run on the core devices that ship with PennyLane. 
+  
+  * If `--shots` is not given, a default of 50000 is used
```
```suggestion
  * If `--shots` is not given, a default of 50000 is used.
```

mariaschuld commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,23 @@
+  
+  * If `--shots` is not given, a default of 50000 is used.
+  
+  * If `--analytic` is not given, the device's default is used
```
```suggestion
  * If `--analytic` is not given, the device's default is used.
```

mariaschuld commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,23 @@
+  
+  * If `--analytic` is not given, the device's default is used.
+                                                                                                                                                        
+  If the test is run on external devices, the device and its dependencies must be 
```
```suggestion
  If the tests are run on external devices, the device and its dependencies must be 
```

mariaschuld commented on pennylane/plugins/tests/__init__.py:
```diff
@@ -0,0 +1,37 @@
+(where ``qiskit.aer`` is replaced by the device to be tested).
+
+Most tests query the device's capabilities and only get executed if they apply to the device.
+Both analytic (with an exact probability distribution) and non-analytic devices (with an estimated
```
```suggestion
Both analytic devices (producing an exact probability distribution) and non-analytic devices (producing an estimated
```

mariaschuld commented on pennylane/plugins/tests/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+>>> python3 -m pytest tests/* --device qiskit.aer
+
+(where ``qiskit.aer`` is replaced by the device to be tested).
```
```suggestion
where ``qiskit.aer`` is replaced by the device to be tested.
```

mariaschuld commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+    return _tol
+
+
+@pytest.fixture(scope="session")
```
I know we want to keep fixtures at a minimum, but I find this one clutters the code if repeated in every function.

mariaschuld commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+    return _init_state
+
+
+@pytest.fixture(scope="session")
```
This is the fixture used to decide whether the device should be tested...

mariaschuld commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+    """Add command line option to pytest."""
+
+    # TODO: find a way to pass an unknown number of optional command line options to pytest
+    # so we can optionally provide compulsory args like "cutoff_dim" or specific args such as a backend
```
I start to believe that this is impossible with pytest - in which case we'd need a useful Plan B!

josh146 replied:
Perhaps it can be done via `getgroup()` or `parse_known_and_unknown_args`? It looks like [PyTest simply wraps argparse](https://docs.pytest.org/en/stable/reference.html#parser), so it should be possible

mariaschuld replied:
Update: unknown args is an extremely hacky solution for argparse it seems. We decided to specify all arguments in the ``config.toml`` file, which is what it was made for.

mariaschuld commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+            device_kwargs["analytic"] = False
+        else:
+            device_kwargs["analytic"] = True
+
```
```suggestion

```

mariaschuld commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,136 @@
+        # use default value of device
+        device_kwargs.pop("analytic")
+    else:
+        # turn string into boolean
```
Not sure if there is a more elegant way, but we cannot use ``type=bool``, because we need a third option ``None`` for if the flag is unused and the default value should be used...

josh146 replied:
Ah 🤔 

In the meantime, might be good to use `device_kwargs["analytic"].lower()` to allow if the user passes `'false'`, to avoid it accidentally becoming True.

mariaschuld commented:
Hey @thisac @antalszava @josh146 @co9olguy, I tagged you for a pre-review. Usually I'd try to minimize work load like this, but I realised that I had to take lots of design decisions that I would like your opinion on. 

This test suite will also determine how devices are written (and may require some updates of plugins), so I think we need a lot of eyes on it... Thanks :heart_decoration:

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,136 @@
+        if analytic:
+            return float(os.environ.get("TOL", TOL))
+        else:
+            return TOL_STOCHASTIC
```
This is a pattern that Nathan brought up with me earlier for not liking (I think I originally wrote a fixture like this for the Qiskit plugin), but I have no idea where the GitHub comment is, I tried to search for it 😆 

I think the original comment was something along the lines of:

1. not liking that the exact tests and the sampled tests had different tolerances

2. sampled tests should not be tested in a way that the stochasticity presents itself

For (2), I don't think this can be bypassed, since these are integration tests, and we don't have the ability to monkeypatch and avoid stochasticity. For (1), I'm at a bit of a loss over how to improve on it.

mariaschuld replied:
I think we should wait for his input, I don't see how to circumvent it (and also not why).

josh146 replied:
:+1: It might also not be an issue any more, since this is clearly an integration test (as opposed to the previous test suite, which 'masqueraded' as a unit test).

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,136 @@
+    # TODO: find a way to pass an unknown number of optional command line options to pytest
+    # so we can optionally provide compulsory args like "cutoff_dim" or specific args such as a backend
+
+    parser.addoption("--device", action="store", default=None)
```
Perhaps we could also add `action='append', nargs='+'` here? This way, `opt.device` could be a list of devices.

josh146 replied:
It would need some surgery in the `pytest_generate_tests` to automatically parametrize all tests using the `device` list.

mariaschuld replied:
That's good to know, but the reason why I don't want to allow for multiple devices is that things get very complicated with the kwargs then, i.e. which kwarg is used for which device...maybe leave that for another time once the basic structure works?

josh146 replied:
Good point!

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,136 @@
+
+            metafunc.parametrize("device_kwargs", list_device_kwargs)
+        else:
+            metafunc.parametrize("device_kwargs", [device_kwargs])
```
Small thought - I noticed below that there are some `skip_if` conditions that are repeated a lot. We could turn these into mark decorators, like we do in SF:

```python
@pytest.mark.backends("gaussian", "fock")
class TestSomething:
```

The nice thing about marks is you can apply them to classes rather than each test, and you can even apply them to an entire file:

```python
mark = pytest.mark.backends("tf")
```

To do this in SF, we define the `pytest_runtest_setup()` function, which is executed prior to each test.

```python
def pytest_runtest_setup(item):
    """Automatically skip tests if they are marked for only certain backends"""
    if tf_available:
        allowed_backends = {"gaussian", "tf", "fock"}
    else:
        allowed_backends = {"gaussian", "fock"}

    # load the marker specifying what the backend is
    marks = {mark.name for mark in item.iter_markers() if mark.name in allowed_backends}

    # load a marker specifying whether the test only works with certain backends
    test_backends = [mark.args for mark in item.iter_markers(name="backends")]

    if not test_backends:
        # if the test hasn't specified that it runs on particular backends,
        # assume it will work with all backends
        test_backends = allowed_backends
    else:
        # otherwise, extract the set of backend strings the test works with
        test_backends = test_backends[0]

    for b in marks:
        if b not in test_backends:
            pytest.skip(
                "\nTest {} only runs with {} backend(s), "
                "but {} backend provided".format(item.nodeid, test_backends, b)
            )

    # skip broken tests
    for mark in item.iter_markers(name="broken"):
        if mark.args:
            pytest.skip("Broken test skipped: {}".format(*mark.args))
        else:
            pytest.skip("Test skipped as corresponding code base is currently broken!")
```
(I don't think ours needs to look this complicated).

josh146 replied:
This way, we could have
```python
@pytest.mark.model("qubit")
```

mariaschuld replied:
That is really cool. But it is again a bit more opaque. Skip_if will tell every coder what the test is run for, while for "mark" there is magic happening...Is this a case of pytest genius versus clarity?  Or do you think it would be more readable with the `mark` decorator?

josh146 replied:
Oh, I see `mark` used quite often in other Python packages, so it feels relatively 'normal' to me.

antalszava replied:
Instead of programmatically separating qubit and cv cases, could we create separate folders for each set of test files (e.g. `tests/qubit` & `tests/cv`)? And files added in this PR would mainly go into `tests/qubit`.

It would feel natural to me that we have the branching logic earlier in the line of execution. This would also help with the organization of the files. (Still picking up details of how the cli for pytest really propagates the info on which test cases to call, so let me know if this would be difficult to achieve)

I'd imagine that actually creating two separate test suites might also not be an entirely radical idea. There would be code duplication involved, but I'd mostly think we are either working on the qubit based parts or the CV based parts. Exceptions could include extending e.g. shared parts of nodes, but we'd have to add different integration tests anyways.

mariaschuld replied:
I'm in favour of that @antalszava, since CV/Qubit only share very few tests anyways, and may require different base settings for the stochastic tests...

josh146 commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+        res = circuit()
+        expected = np.abs(rnd_state) ** 2
+
+        assert np.allclose(res, expected, atol=tol(dev.analytic))
```
Can we have the tolerance fixture depend on the `device` fixture, to automatically get the `analytic` attribute value?

mariaschuld replied:
That would be cool, but I don't see how it works, since the device is only created in the test. One could use `device_kwargs` but it only contains a specific value for `analytic` if this was set from command line (else it is None and will be defined once the device is created). 

In short: it is only defined inside a test whether the test is run analytic or not.

josh146 commented:
Also, re: codefactor and codecov:

* With codefactor, we currently don't lint the test folder. We could either set up `pennylane/plugins/tests` to also be skipped, or lint the test folder.

* I assume codecov is failing because the new tests directory counts as PennyLane source code, and since these tests aren't being run on Travis, they aren't being marked as covered.

mariaschuld commented:
Thanks for your review Josh! Very helpful. I will try to crack the pytest unknown arguments somehow, but it's poorly documented. 

- About code factor: I don't mind skipping or linting the folder. What would you prefer? Where could I set this?
- About codecov: I added an "ignore" flag to `codecov.yml`.

I think the stochasticity is a deeper discussion we should have with @co9olguy .

mariaschuld commented:
By the way, I already used the test when developing the qulacs plugin and they are very helpful in picking up capabilities a device does not have! I think this could become very useful.

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+   
+  * If `--device` is not given, the tests are run on the core devices that ship with PennyLane. 
+  
+  * If `--shots` is not given, a default of 50000 is used.
```
Would 50000 be a bit too many for default? :thinking: We have a 1000 as the default for the device creation. Or could it make sense to have a separate default number of shots for analytic and non-analytic cases?

josh146 replied:
Oh, this is a good point @antalszava --- Qiskit has a maximum of 8192 shots as well!

mariaschuld replied:
Cool, let me set 10,000

antalszava replied:
@mariaschuld just double-checking: would it be 10,000 or 1,000 here?

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+  
+  * If `--analytic` is not given, the device's default is used.
+  
+  Other arguments of the device, such as `strawberryfields.fock`'s compulsory `cutoff_dim`, 
```
(Minor suggestion): just to have a qubit model example for this CHANGELOG entry (as CV would be upcoming)
```suggestion
  Other arguments of the device, such as `qiskit.aer`'s compulsory `backend_options`, 
```

antalszava commented on .gitignore:
```diff
@@ -29,3 +29,4 @@ timer.dat
 tmp/*
 benchmark/revisions/
 venv
+config.toml
```
:100:

antalszava commented on codecov.yml:
```diff
@@ -1,2 +1,5 @@
   - "pennylane_qchem/::qchem/pennylane_qchem/"
+
+ignore:
+  - "pennylane/plugins/tests/*"
```
Hmm, I wonder how coverage is going to be affected by having the shared test suite. Mostly for plugins :thinking: Might be that we'll have to think of a "shared coverage" option?

josh146 replied:
Once we have the CI executing the shared tests, the coverage will be fine 🙂 With tests, all you need to do is execute them and they'll be counted for full coverage (and `default.gaussian`, `default.qubit`, `default.qubit.tf` should be able to execute all tests)

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+
+# list of all parametrized single-qubit gates
+single_qubit_param = [
+    # (qml.PhaseShift(0, wires=0), phase_shift),
```
?

mariaschuld replied:
Anachronism...nice find!

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+    (qml.T, T),
+]
+
+# list of all parametrized single-qubit gates
```
(Minor suggestion): e.g. `Rot` is not included
```suggestion
# list of all parametrized single-qubit gates taking a single parameter
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+"""
+import pytest
+
+import numpy as np
```
for later uses
```suggestion
import math
import cmath
import numpy as np
```

mariaschuld replied:
Thanks, didn't know that one before!

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+
+# parametrized qubit gates
+phase_shift = lambda phi: np.array([[1, 0], [0, np.exp(1j * phi)]])
+rx = lambda theta: np.cos(theta / 2) * I + 1j * np.sin(-theta / 2) * X
```
```suggestion
rx = lambda theta: math.cos(theta / 2) * I + 1j * math.sin(-theta / 2) * X
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+        [0, 0, 0, np.exp(1j * theta / 2)],
+    ]
+)
+
```
(blacking may be needed)
```suggestion
cry = lambda theta: np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, math.cos(theta / 2), -math.sin(theta / 2)], [0, 0, math.sin(theta / 2), math.cos(theta / 2)]])
crx = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, math.cos(theta / 2), 1j * math.sin(-theta / 2)], [0, 0, 1j * math.sin(-theta / 2), math.cos(theta / 2)]])

```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+# list of all non-parametrized two-qubit gates
+two_qubit = [(qml.CNOT, CNOT), (qml.SWAP, SWAP), (qml.CZ, CZ)]
+# list of all parametrized two-qubit gates
+two_qubit_param = [(qml.CRZ, crz)]
```
```suggestion
two_qubit_param = [(qml.CRX, crx), (qml.CRY, cry), (qml.CRZ, crz)]
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+    @pytest.mark.parametrize("theta", [0.5432, -0.232])
+    @pytest.mark.parametrize("op,func", two_qubit_param)
+    def test_two_qubit_parameters(self, device, init_state, op, func, theta, tol, skip_if):
+        """Test parametrized two qubit gates."""
```
```suggestion
        """Test parametrized two qubit gates taking a single argument."""
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+    @pytest.mark.parametrize("theta", [0.5432, -0.232])
+    @pytest.mark.parametrize("op,func", single_qubit_param)
+    def test_single_qubit_parameters(self, device, init_state, op, func, theta, tol, skip_if):
+        """Test single qubit gates."""
```
```suggestion
        """Test single qubit gates taking a single scalar argument."""
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+
+    @pytest.mark.parametrize("op,mat", single_qubit)
+    def test_single_qubit_no_parameters(self, device, init_state, op, mat, tol, skip_if):
+        """Test inverse PauliX application."""
```
```suggestion
        """Test inverse single qubit gate application."""
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+    @pytest.mark.parametrize("theta", [0.5432, -0.232])
+    @pytest.mark.parametrize("op,func", single_qubit_param)
+    def test_single_qubit_parameters(self, device, init_state, op, func, theta, tol, skip_if):
+        """Test inverse single qubit gates."""
```
```suggestion
        """Test inverse single qubit gates each taking one scalar parameter."""
```

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+        dev = device(n_wires)
+        capabilities = dev.__class__.capabilities()
+        skip_if("model" not in capabilities or not capabilities["model"] == "qubit")
+        skip_if("inverse_operations" not in capabilities)
```
Should there be some logging done in such cases to inform that these tests were skipped?

antalszava commented on pennylane/plugins/tests/test_gates.py:
```diff
@@ -0,0 +1,489 @@
+    @pytest.mark.parametrize("theta", [0.5432, -0.232])
+    @pytest.mark.parametrize("op,func", two_qubit_param)
+    def test_two_qubit_parameters(self, device, init_state, op, func, theta, tol, skip_if):
+        """Test inverse parametrized two qubit gates."""
```
```suggestion
        """Test inverse of two qubit gates taking one parameter."""
```

antalszava commented on pennylane/plugins/tests/test_properties.py:
```diff
@@ -0,0 +1,47 @@
+    """Test the device is created with the expected properties."""
+
+    def test_load_device(self, device_kwargs):
+        """Test that the QVM device loads correctly."""
```
```suggestion
        """Test that the device loads correctly."""
```

mariaschuld commented:
Thanks so much for your code review @antalszava and @josh146 .

Overall I implemented your very useful suggestions where ever I could:

* Added the new gates like CRY to tests
* Make pytest exit if device cannot be created
* Lowered default shots to 10,000
* fixed complaints from codefactor
* Use math functions instead of numpy
* Ignore tests in coverage report (we can change this if we decide to run the tests by CI, in which case they are covered)

A few things were tricky and I suggest we revisit them with a refactor of the devices themselves:

* _Splitting cv and qubit tests_ - For now the tests will be qubit only. The ``default.gaussian`` device is removed from the list of automatically tested core devices, and the tests will exit if the model is not qubit.
* _Use ``pytest.mark`` to skip tests_ - I tried to implement that, and it would be much neater. We could just parametrize a test with ``@pytest.mark.capabilities("inverse_operations.true", "model.qubit")`` or so. But it seems impossible to retrieve the device capabilities in the hook, and I cannot find an option to make it work. So I cleaned up the ``skip_if`` fixture a bit, and we can revisit this in future. At the moment, `mark` would not reduce code.
* _Use ``slow`` decorator to mark stochastic tests_ - if the user chooses a non-analytic device, this would skip all tests but 2 :) But since the stochastic tests run quite fast, I think this is not necessary?
* _Making the ``tol`` fixture use the device fixture to decide whether to represent the analytic or non-analytic tolerance_ - the device is only created inside the test, since the test only knows the correct number of wires...This is a basic design decision that we could change (i.e., hack the correct wires into an instance of the device later), but I don't know how stable this is for the future.

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
 
 <h3>New features since last release</h3>
 
+* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 
```
@mariaschuld ever since you mentioned it, I definitely thing we should rename `pennylane.plugins` to `pennylane.devices`

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 
+  to run generic tests on core or external devices calling 
+  
+  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                    
```
🎉

josh146 replied:
I did some research on this, and it looks like this is the exact solution we eventually want: https://stackoverflow.com/questions/39091443/running-py-test-tests-from-another-pip-installed-package-against-local-code:

> I am building a package that provides (1) a fixed interface with (2) pluggable drivers. The desired driver will be selectable by the user of the package. My package will include at least one driver with it, but I would like other developers to be able to implement and validate drivers that conform to the package's interface. As such, I would like these developers to be able to run my tests against their drivers.

---

As a short term solution, we could:

* Add the function `qml.get_device_tests()` that returns the test location on the users system (similar to [`np.get_include()`](https://numpy.org/devdocs/reference/generated/numpy.get_include.html).

* Provide a convenience CLI script alongside PennyLane, that is simply a wrapper around `pytest qml.get_device_tests()`.

mariaschuld replied:
I added this to the list of next milestones!

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,25 @@
+  
+  * If `--analytic` is not given, the device's default is used.
+  
+  Other arguments of the device, such as `qiskit.aer`'s compulsory `backend_options`, 
+  can be defined in the `config.toml` file containing custom PennyLane configurations.
```
:+1:

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -40,6 +59,7 @@ Antal Száva, Nicola Vitucci
 <h4>New and improved simulators</h4>
 
 * Added a new device, `default.qubit.tf`, a pure-state qubit simulator written using TensorFlow.
+
```
```suggestion
```

josh146 commented on pennylane/plugins/tests/__init__.py:
```diff
@@ -0,0 +1,40 @@
+This subpackage provides integration tests for the devices with PennyLane's core
+functionalities. At the moment, the tests only run on devices based on the 'qubit' model.
+
+To run the tests, navigate to the parent directory of this file and run (i.e., for 'default.qubit'):
+
+>>> python3 -m pytest tests/* --device default.qubit --shots 1234 --analytic False
```
(note: `>>>` is reserved for python code)

```suggestion
To run the tests against a particular device (i.e., for 'default.qubit'):

.. code-block:: console

    python3 -m pytest path_to_pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False

The location of your PennyLane installation may differ depending on installation method and operating
system. To find the location, you can execute the following Python code:

>>> import os
>>> import pennylane as qml
>>> print(os.path.dirname(qml.__file__))
```

josh146 replied:
(the last three lines of python could be provided as a convenience function, e.g., `pennylane.get_device_tests()` or something along those lines)

josh146 commented on pennylane/plugins/tests/__init__.py:
```diff
@@ -0,0 +1,40 @@
+
+>>> python3 -m pytest tests/* --device default.qubit --shots 1234 --analytic False
+
+The command line arguments are optional.
```
```suggestion
The command line arguments are optional:
```

josh146 commented on pennylane/plugins/tests/__init__.py:
```diff
@@ -0,0 +1,40 @@
+
+* If `--device` is not given, the tests are run on the qubit core devices that ship with PennyLane.
+
+* If `--shots` is not given, a default of 50000 is used.
```
Does the 50000 need to be modified?

josh146 replied:
might be worth specifying that this option does not apply if `analytic=True`

josh146 replied:
We need a better term than 'analytic', don't we?

mariaschuld replied:
Yes!! But I don't have a better one either...

Yes, I thought I changed the 50000 to 10000, but something must have merged in.

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+# Number of shots to call the devices with
+N_SHOTS = 10000
+# List of all devices that are included in PennyLane
+LIST_CORE_DEVICES = {"default.qubit", "default.qubit.tf"}
```
Curios for the reason for leaving out `default.tensor`?

mariaschuld replied:
I wasn't sure if the beta devices are strictly speaking core devices?

mariaschuld replied:
I tried to add default.tensor and default.tensor.tf, and it turns out they do not have an "analytic" attribute. 

This is one of those issues with the PR - the tests are supposed to work for the existing devices, but at the same time they should define what a device must be able to do. I would vote for making "analytic" a compulsory attribute (or capability), since everything hinges on it and leaving the tests as is for now?

josh146 replied:
Oh, good point. As far as I know, both devices are (by default) `analytic=True`, they don't support finite sampling. They should really define the attribute --- I agree with making it mandatory.

co9olguy replied:
The reason these devices don't have an `analytic` attribute is simply because they do not inherit from `QubitDevice`. Doesn't mean they aren't actually "analytic" though :) 

I feel this is something that is better declared in `_capabilities` though

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+    def _skip_if(dev, capabilities):
+        """ Skip test if device has any of the given capabilities. """
+
+        dev_capabilities = dev.__class__.capabilities()
```
How come you have to call `dev.__class__`? Shouldn't the instantiated device also have the same capabilities method?

mariaschuld replied:
Oh true. I thought you need to call a class method on the class, not the instance...

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+
+    def _tol(analytic):
+        if analytic:
+            return float(os.environ.get("TOL", TOL))
+        else:
+            return TOL_STOCHASTIC
```
```suggestion
            return float(os.environ.get("TOL", TOL))
        return TOL_STOCHASTIC

```

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+LIST_CORE_DEVICES = {"default.qubit", "default.qubit.tf"}
+
+
+@pytest.fixture(scope="function")
+def tol():
```
Minor suggestion, but pytest allows for fixture injection. So for instance, you could do the following:

```python
@pytest.fixture(scope="function")
def tol(analytic):
    if analytic:
        return {"atol": float(os.environ.get("TOL", TOL)), "rtol": 0}
    return {"atol": TOL_STOCHASTIC, "rtol": 0}
```

where the `analytic` fixture is _not yet defined_.

When running the test, you simply declare the value of the analytic fixture using `@pytest.mark.parametrize`:

```python
@pytest.mark.parametrize(analytic, [True]):
def test_something(tol):
    ...
    assert np.allclose(result, expected, **tol)
```
the test doesn't even need to have `anlaytic` in the signature, it is injected directly into any dependent fixtures (in this case, `tol`).

You can also run the test for multiple values as usual:


```python
@pytest.mark.parametrize(analytic, [True]):
def test_something(analytic, tol):
    dev = qml.device("dsfasd", analytic=analytic)
    assert np.allclose(result, expected, **tol)
```

josh146 replied:
We already use this pattern in the interface tests (fixtures have implicit dependence on an undefined `interface` fixture, which is then defined at the module/class/function level using `parametrize`). The main simplification of this approach is
* simplification of test signatures
* simplification of fixtures (no need to make them factories).

This comes at the expense that developers need to remember that `tol` requires that `analytic` be defined, even for a single value.

mariaschuld replied:
I really like it (and it makes me wonder if I chose the right design in the first place), but "analytic" is not really a fixture chosen in a test, but a property of the device which is only created in the test itself...

I don't see how I can make use of fixture injection here?

josh146 replied:
🤔

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+            dev = qml.device(**device_kwargs)
+        except qml.DeviceError:
+            # exit the tests if the device cannot be created
+            pytest.exit(
+                "Device {} cannot be created. To run the device tests on an external device, the "
+                "plugin and all of its dependencies must be installed.".format(
+                    device_kwargs["name"]
```
nice!

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+                )
+            )
+
+        capabilities = dev.__class__.capabilities()
+        if "model" not in capabilities or not capabilities["model"] == "qubit":
+            # exit the tests if device based on cv model (currently not supported)
+            pytest.exit("The device test suite currently only runs on qubit-based devices.")
```
:+1:

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+                )
+            )
+
+        capabilities = dev.__class__.capabilities()
```
same question here re: `__class__`

josh146 commented on pennylane/plugins/tests/conftest.py:
```diff
@@ -0,0 +1,177 @@
+
+    # skip tests marked as broken
+    for mark in item.iter_markers(name="broken"):
+        if mark.args:
+            pytest.skip("Broken test skipped: {}".format(*mark.args))
+        else:
+            pytest.skip("Test skipped as corresponding code base is currently broken!")
```
Nice 💯 This is a great way of keeping track of broken tests!

josh146 commented on pennylane/plugins/tests/test_measurements.py:
```diff
@@ -0,0 +1,728 @@
+
+        res = circuit()
+
+        # res should only contain the eigenvalues of
+        # the hermitian matrix
+        eigvals = np.linalg.eigvalsh(A_)
+        assert np.allclose(sorted(list(set(res))), sorted(eigvals), atol=tol(dev.analytic))
+        # the analytic mean is 2*sin(theta)+0.5*cos(theta)+0.5
+        assert np.allclose(
+            np.mean(res), 2 * np.sin(theta) + 0.5 * np.cos(theta) + 0.5, atol=tol(False)
+        )
+        # the analytic variance is 0.25*(sin(theta)-4*cos(theta))^2
+        assert np.allclose(
+            np.var(res), 0.25 * (np.sin(theta) - 4 * np.cos(theta)) ** 2, atol=tol(False)
+        )
```
I've always been unhappy with the way these sample tests are designed; essentially, they are checking that

* The returned samples have the allowed values
* The returned samples have the right statistical properties up to a tolerance 

Do you think this is sufficient? I've never been able to come up with a better end-to-end integration tests for sampling. (If this was a unittest, and we had access to the simulator code, we would be able to simply monkeypatch and check if the deterministic quantities used to generate the samples were correct...)

mariaschuld replied:
It's our old question. I think we should set a standard for ourselves!

co9olguy replied:
I'm not sure there's much else to test :thinking:. You could add more random variables, by the mean and variance seem like the natural things to check
